test/hotspot/jtreg/runtime/logging/LoaderConstraintsTest.java

```diff
@@ -49,22 +48,23 @@ public class LoaderConstraintsTest {
     private static ProcessBuilder pb;
     private static class ClassUnloadTestMain {
         public static void main(String... args) throws Exception {
-            String className = "test.Empty";
             ClassLoader cl = ClassUnloadCommon.newClassLoader();
-            Class<?> c = cl.loadClass(className);
-            cl = null; c = null;
-            ClassUnloadCommon.triggerUnloading();
+            Class<?> c = cl.loadClass("test.Empty");
+            // Causes class test.Emtpy to be linked, which triggers the
+            // constraint on class String due to override of toString().
+            Constructor<?> constructor = c.getDeclaredConstructor();
         }
     }
 
     // Use the same command-line heap size setting as ../ClassUnload/UnloadTest.java
     static ProcessBuilder exec(String... args) throws Exception {
+        String classPath = System.getProperty("test.class.path", ".");
+
         List<String> argsList = new ArrayList<>();
         Collections.addAll(argsList, args);
         Collections.addAll(argsList, "-Xmn8m");
         Collections.addAll(argsList, "-Xbootclasspath/a:.");
```

Comment:
Without use of `WhiteBox`, you don't need this. And this is probably what caused the class to be loaded by the boot-loader instead of the custom loader as expected. But the use of WB is needed by `ClassUnloadCommon` and these instructions are present to ensure it works - as added by JDK-8289184.

Reply:
Added back the use of `WhiteBox`. As to `test.Empty`, it was always loaded by the `app` classloader regardless of `-Xbootclasspath/a:.`, i.e that’s only needed for loading `WhiteBox`. The issue was the missing `test.class.path` property. Class `test.Empty` resides in directory `jtreg_test_hotspot_jtreg_runtime_logging_LoaderConstraintsTest_java/classes/0/runtime/logging/classes/test/` but the process’s current working directory is `jtreg_test_hotspot_jtreg_runtime_logging_LoaderConstraintsTest_java/scratch/0`, so setting the custom classloader classpath as `.` meant it wasn’t able to find it. So we ended up delegating to the parent classloader (`app`) which already had the correct directories in the classpath.